### ISSUE TITLE
[Vue v3] Changes for @Component decorator and Vue base class

### ISSUE THREAD

**Issue by ktsn** [vuejs/vue-class-component#406]
## Summary

- `@Component` will be renamed to `@Options`.
- `@Options` is optional if you don't declare any options with it.
- `Vue` constructor is provided from `vue-class-component` package.
- `Component.registerHooks` will move to `Vue.registerHooks`

Example:

```vue
<template>
  <div>{{ count }}</div>
  <button @click="increment">+1</button>
</template>

<script>
import { Vue, Options } from 'vue-class-component'

// Component definition
@Options({
  // Define component options
  watch: {
    count: value => {
      console.log(value)
    }
  }
})
export default class Counter extends Vue {
  // The behavior in class is the same as the current
  count = 0

  increment() {
    this.count++
  }
}
</script>
```

```js
// Adding lifecycle hooks
import { Vue } from 'vue-class-component'

Vue.registerHooks([
  'beforeRouteEnter',
  'beforeRouteLeave',
  'beforeRouteUpdate'
])
```

## Details

As Vue v3 no longer provides base `Vue` constructor, Vue Class Component will provide it instead. Fortunately, we can add class component specific features in the base class as we define it in Vue Class Component package.

One of the benefits with the new `Vue` constructor is we can make `@Component` decorator optional. There were non-trivial number of confusions regarding missing `@Component` decorator in super class / mixins (e.g. #180). Making decorator optional would solve this problem.

Also renaming `@Component` with `@Options` would make more sense as it is for adding component options to your class components rather than making a class component.

Since `@Options` is optional, having [`registerHooks`](https://class-component.vuejs.org/guide/additional-hooks.html) on the decorator will not work. So we will move the method under static field of `Vue` constructor.

### Declaring component options as static properties

We could do this to define class component options:

```js
export default class Counter extends Vue {
  // All component options are static properties
  static watch = {
    count: value => {
      console.log(value)
    }
  }

  count = 0

  increment() {
    this.count++
  }
}
```

But it has a drawback: we cannot define static properties / methods not for component options both on userland and library side. e.g. we cannot define static method like `registerHooks` on `Vue`.

**Comment by sodatea** [comment]
Seems some of the breaking changes may not be fully covered with a codemod.
Is it possible to implement a runtime compatibility layer on top of it in `vue-property-decorator`?

**Comment by ktsn** [comment]
Yes, we can provide a runtime compatibility layer. Just wrapping `Options` with `Component` as `Options` provides almost the same API interface and feature with the current `Component.`

Maybe providing `Options` and `Vue` from `vue-class-component` with minor update would help to migrate.

**Comment by nicolidin** [comment]
Is there a way to already test it with the vuejs/composition-api ??
Thanks!

**Comment by ktsn** [comment]
I'm going to create a bridge api for composition functions in a separated issue.

**Comment by the-emerald** [comment]
Does the abandonment of the [Class API proposal](https://github.com/vuejs/rfcs/pull/17) have any affect on this package and whether it would be maintained at the same level in the future? Cheers.

**Comment by HIMISOCOOL** [comment]
> Does the abandonment of the [Class API proposal](https://github.com/vuejs/rfcs/pull/17) have any affect on this package and whether it would be maintained at the same level in the future? Cheers.

nope, from what I understand this package will always exist and will always have 1st party package status, its just not practical for the base vue package to use classes in todays world.

**Comment by 734204679** [comment]
pls update the official documents
https://class-component.vuejs.org/guide/additional-hooks.html

**Comment by LifeIsStrange** [comment]
@ktsn So we did try the latest vue class component beta and while the new props declaration is working just fine, we need to have Watch, Provide and other related Vue features.
Putting watch in @Options works but is not expressive enough at all. The static alternative is not yet implemented but would suffer the same issue being static ?

The issue is that we were able to access the class variables through `this` inside the method that react to the thing it watch.
This expressivity is *needed* for a lot of behavior and we are unable to migrate to Vue3 until this crucial limitation is fixed.

**Comment by calebeaires** [comment]
@LifeIsStrange like you, our team share with those concerns. Waiting for an update!

**Comment by iklemm** [comment]
Seems that this new API does not match with documentation https://class-component.vuejs.org which is very confusing.

**Comment by Chris2011** [comment]
@LifeIsStrange I think having Watch and Provide and other decorators for class methods and fields are part of vue-property-decorators, isn't it?

**Comment by LifeIsStrange** [comment]
@Chris2011
Vue property decorator has not be ported to vue 3 and nobody is working on it for the foreseeable future. This is very worrying for the industry.

**Comment by Chris2011** [comment]
@LifeIsStrange well, vue-property-decorator depends on vue-class-component, so I guess they will farely wait for the new version 8 of this package here. :). Just my opinion, no insides.

**Comment by calebeaires** [comment]
Hope this situation changes. This two libs belongs to the top of mind of the Vue community!

**Comment by LifeIsStrange** [comment]
@Chris2011 that's not how it should work, they should begin working on it *while* it's in beta. If they need to do some breaking changes to vue class component it will be too late after a stable release

**Comment by ZhengXingchi** [comment]
how can we modify a msg in watch method like that

![image](https://user-images.githubusercontent.com/31958613/100708517-6d59dd80-33e7-11eb-84dd-2733e08cfa5e.png)

**Comment by ZhengXingchi** [comment]
@ktsn

**Comment by ygj6** [comment]
@ZhengXingchi   this cannot be used in arror functions, the following code is work
watch: {
    count( value) {
      console.log(value)
      if(value==10){
        this.msg='计数超过10了'
      }
    }
  }

**Comment by theoephraim** [comment]
Maybe it's time that vue-class-component and vue-property-decorator merge into a single tool? It seems like it would be easier to design a better and more wholistic solution this way.

**Comment by nicolidin** [comment]
I totally agree with @theoephraim, I think it would be better and more pragmatic for the class component ecosystem to merge these two libraries. 
I also think it is the wish of many users of this library.

**Comment by calebeaires** [comment]
I am with you. I use class component style to my project. Don't migrate to Vue 3 until this movement is mature!

**Comment by aliibrahimroshan** [comment]
I am migrating to vue 3 . I am not be able to access the props value in my class 

import { Options, Vue } from "vue-class-component";


@Options({
  props: {
    titleComponent: Number
  }
})

export default class StorePin extends Vue {

    private titleComponent = 0;
  
  };

  mounted()
  {
    console.log(this.propValue);
  }

  get propValue()
  {
    return this.titleComponent;
  }

  incremnt()
  {
    this.titleComponent ++ ;
  }

}
</script>

**Comment by zischler** [comment]
@aliibrahimroshan If you didn't find the answer yet: See #465 
I've done it like that and not in the `@Options`
```
class Props {
  // optional prop
  foo?: string

  // required prop
  bar!: string

  // optional prop with default
  baz = prop<string>({ default: 'default value' })
}

export default class MyComp extends Vue.with(Props) {}
```

**Comment by ly1214** [comment]
I use class component style to my project.
 filter 

`@Options({
  // props: {
  //   docList: Array
  // },
  filters: {
    formatTime: (time: string) => {
      console.log('time', time)
    }
  },
  watch: {
    count: value => {
      console.log(value)
    }
  }
})`
[Vue warn]: Property "filterName" was accessed during render but is not defined on instance


![2222222](https://user-images.githubusercontent.com/37330588/103870286-75075300-5106-11eb-886a-59440ef3004e.png)

**Comment by bhabgs** [comment]
use vue-class-component in `.tsx`. 
* render function 
![image](https://user-images.githubusercontent.com/19221378/104678774-2752a200-5727-11eb-9086-aac961269449.png)

`I have to set it to any`

**Comment by hsbtr** [comment]
The new API is not reflected in the document https://class-component.vuejs.org/, and this document is a little rough, I hope to improve the document, if it will be maintained in the future, it will be better to put it in the VUE3 document

**Comment by websmurf** [comment]
For those running into issues with Vue3 and vue-property-decorator; have a look at this package: https://github.com/calebeaires/vue-decorator

That solved most of my issues.

**Comment by Chris2011** [comment]
> For those running into issues with Vue3 and vue-property-decorator; have a look at this package: https://github.com/calebeaires/vue-decorator
> 
> That solved most of my issues.

Thats cool and also to add it here, wouldn't it be better to make a PR to this project here @calebeaires ?

**Comment by tsofist** [comment]
Can you tell me?🎈
Are there any significant changes planned in v8 compared to the current v8-rc1

**Comment by softboy99** [comment]
> I totally agree with @theoephraim, I think it would be better and more pragmatic for the class component ecosystem to merge these two libraries.
> I also think it is the wish of many users of this library.

I agree with you @theoephraim ,nicolidin   too

**Comment by nicolidin** [comment]
> how can we modify a msg in watch method like that

@ZhengXingchi 
You can use `watch` through `this.$watch` like bellow:

``` typescript
count = 0;

created() {
  this.$watch("count", () => {
    console.log(count updated);
  });
}
```

**Comment by fobdy** [comment]
@ktsn Hey guys. Thank you for your hard work.
One question from me: will it be possible to make the `v8` work within `vue-2` + `vue/composition-api` plugin environment?

It's pretty critical for us since we're going to smoothly transition from vue-2 to vue-3 that's why we decided to start touching the composition-api a little bit within the current codebase. Also, we've enabled some of the recommended vue3 es-lint rules to help us transition a bit since no vue-3 "migration build" is provided yet (and nobody knows when it happens).

**Comment by Exotelis** [comment]
It seems that @kaorun343 is working on an upgrade of `vue-property-decorator` to make it compatible with vue3.

The question has been asked a couple of times already: Wouldn't it make sense to merge both projects together? Personally, I'd prefer using one instead of two packages. Unfortunately, I haven't seen a statement from any of the package maintainers. I'm upgrading one of my projects to vue3 and I'd love to know if I should use `vue-property-decorator` again or if I should wait for `vue-class-component` to have the same set of features.

Apart from that: Getting started with `vue-class-component v8.0.0-rc.1` is kinda tricky without any kind of documentation. 

Thank you for your great work :)

**Comment by doronguttman** [comment]
in case someone gets here looking for how to use `setup` functionality with the current solution (as of [v8.0.0-rc1](https://github.com/vuejs/vue-class-component/releases/tag/v8.0.0-rc.1)), here's what I was able to figure out:
1. create your class component, e.g.:
   ```typescript
   import { Vue } from "vue-class-component";
   
   export class MyComp extends Vue {
   
   }
   ```
1. add a data field which requires `setup`, e.g.:
   ```typescript
   import { Vue } from "vue-class-component";
   
   export class MyComp extends Vue {
     private someData = "some data";
   }
   ```
1. add the `setup` import and use it as a field factory, e.g.:
    ```typescript
    import { Vue, setup } from "vue-class-component";
    
    export class MyComp extends Vue {
      private someData = setup(() => "some data");
    }
    ```
1. if you want an async setup, you can do that as well, e.g.:
    ```typescript
    import { Vue, setup } from "vue-class-component";
    
    export class MyComp extends Vue {
      private someData = setup(async () => await someService.getAsync());
    }
    ```

**Comment by Micene09** [comment]
Hi @ktsn , any ETA on the release of the library?

Me and my team still waiting for so long, we know that the Composition API is and will be main stream for a while...but we love the class style syntax.
It is compact, more readable, faster (because of getters and data() abstraction)...and with the setup support now, will be more pluggable with shared composables.

Are there any remaining tasks or todos? I would help if the project need it.

**Comment by turbobuilt** [comment]
@Micene09 you can do this in vue 3

```
npm i vue-class-component@8.0.0-rc.1
```
then in code

```
import { Options, Vue } from "vue-class-component";
import HelloWorld from './components/HelloWorld.vue'

@Options({})
export default class App extends Vue {

}
```

**Comment by bodograumann** [comment]
All my components are anonymous now, do I have to add a manual name attribute?

**Comment by bodograumann** [comment]
It seems after an update to vue-3.1.1 it is not possible to access data attributes on the component anymore.
Can anyone confirm this?

**Comment by dmytropryshliak** [comment]
> It seems after an update to vue-3.1.1 it is not possible to access data attributes on the component anymore.
> Can anyone confirm this?

Yes, have same issue.

**Comment by bodograumann** [comment]
@dmytropryshliak I discovered in the meantime, that the problem only occurs for me, when I update typescript from 4.2 to 4.3. :confused:

**Comment by dmytropryshliak** [comment]
I have typescript 4.3.2 installed on my side.

**Comment by SoftwareDesign-Solution** [comment]
If I omit the constructor, the error message "Uncaught (in promise) TypeError: can't convert undefined to object" does not appear. But I need the constructor because I want to load dependencies here via a container 

`
<template>
  <div>

    <div>{{ value }}</div>

    <button @click="onDebugClicked">Debug</button>

  </div>
</template>

<script lang="ts">
import { Vue, Prop } from 'vue-property-decorator'
//import { container } from "tsyringe";

export default class ClassComponent extends Vue {

  @Prop({ type: String, required: true})
  private value!: string;

  constructor() {
    super();
    console.log('constructor')
  }

  onDebugClicked(): void {
    console.log('onDebugClicked');
  }

}
</script>

<style>

</style>
`

**Installed Packages**
vue 3.0.0.0
vue-class-component 8.0.0.0
vue-property-decorator 10.0.0-rc.3

**Comment by meixger** [comment]
@SoftwareDesign-Solution I got the same error. Resolved the error message by supplying an empty array to the parent constructor:

```
public constructor() {
  super([]);
  console.log('constructor');
}
```

**Comment by AleksandrKoltsov** [comment]
how can i  import component in vue-class-component vue 3? @Options doesn't work, i do this:

<template>
        nameComponent
</template>

import { Options, Vue } from 'vue-class-component';
import nameComponent from './somePath'

@Options({
      components: {
             nameComponent
      }
})
export default class MainComponent extends Vue {}


and i get this result

**Comment by MarsCaiWORD** [comment]
Hi everyone, I have a question. I tried to use Vue3 Extend to build my class base but there will be an error. Can anyone tell me what the problem is?

**BaseComponent.ts**
@Options({})
export default class BaseComponent extends Vue {
}

**MainComponent.ts**
import BaseComponent from 'BaseComponent .ts';
@Options({})
export default class MainComponent extends BaseComponent {
}
**Error**
![image](https://user-images.githubusercontent.com/14837356/132829029-0b09001a-4f84-42c7-9216-a8c3160f2131.png)

**Comment by msebi** [comment]
@ktsn 

Please update the documentation:

https://class-component.vuejs.org/guide/class-component.html#other-options

Came by this post by mistake. It's stuff like this that keeps me (and others from donating).

Thanks!!!!

**Comment by Chris2011** [comment]
@ktsn it would be good if you can enable the "discussions" section to move those discussions to the section where it makes more sense.

**Comment by loeffel-io** [comment]
any update?

**Comment by calebeaires** [comment]
@loeffel-io, waiting as you. maybe the dont have intention to update this lib.

**Comment by JohnBernardsson** [comment]
We are already migrating every class-component to usual Object options API. Hard thing to do, but we needed to take some decision in order to update to Vue 3 ASAP...

I am a bit dissapointed though. Although I appreciate A LOT the work that has been done on this fantastic lib, I would have preferred that the communication about the future of it had been better. I waited months to take the decision pointed above, just because of this lack of clarity.

**Comment by Chris2011** [comment]
> We are already migrating every class-component to usual Object options API. Hard thing to do, but we needed to take some decision in order to update to Vue 3 ASAP...
> 
> I am a bit dissapointed though. Although I appreciate A LOT the work that has been done on this fantastic lib, I would have preferred that the communication about the future of it had been better. I waited months to take the decision pointed above, just because of this lack of clarity.

Why not migrating to the composition api? What was the reason for migrating from class components to the options API?

**Comment by charles-allen** [comment]
I noticed this comment in the [FAQ for Composition API](https://vuejs.org/guide/extras/composition-api-faq.html#relationship-with-class-api):

> Relationship with Class API
>
> We no longer recommend using Class API with Vue 3, given that Composition API provides great TypeScript integration with additional logic reuse and code organization benefits.

I'm sad to see this -- I was a big fan of augmenting vanilla JS (fields, getters, etc) instead of learning a meta language on top -- but it seems that's not the direction the community is going (and for valid advantages). Our team tried refactoring a couple of our components that were written with `vue-class-component` to Composition API, and I was at least pleasantly surprised how easy it was to migrate. I would agree that the TypeScript support is better; we were able to remove a lot of explicit declarations where they are now inferred.

**Comment by Chris2011** [comment]
> I noticed this comment in the [FAQ for Composition API](https://vuejs.org/guide/extras/composition-api-faq.html#relationship-with-class-api):
> 
> > Relationship with Class API
> > We no longer recommend using Class API with Vue 3, given that Composition API provides great TypeScript integration with additional logic reuse and code organization benefits.
> 
> I'm sad to see this -- I was a big fan of augmenting vanilla JS (fields, getters, etc) instead of learning a meta language on top -- but it seems that's not the direction the community is going (and for valid advantages). Our team tried refactoring a couple of our components that were written with `vue-class-component` to Composition API, and I was at least pleasantly surprised how easy it was to migrate. And we did see the benefits of improved TypeScript support; we were able to remove a lot of explicit declarations where they are now inferred.

Plus the benefit of the setup attribute inside of the script tag definition to remove the boilerplate code. So at the end, I recommend to test the composition API. It is very easy and you don't need to use the data: {}, props: {} etc. if you use the setup attribute.

**Comment by Micene09** [comment]
Agree with you guys, Composition API is the right way to go.
It's not easy for me to admit, not just for how much I loved the Class-Style syntax thanks to this lib, but because of the huge amount of code-base that me and my team (in my company) we have done, I'm talking about 5 years of code-base...and I'm pretty sure that there are so many companies stuck with the same problem out there.

The only thing I can suggest here, is even the only way I found to face the problem: [vc2c](https://github.com/yoyo930021/vc2c) (Vue Class 2 Composition) by [yoyo930021](https://github.com/yoyo930021).
It's an auto code-converter from Class Syntax to Composition API, both compatible with Vue 3 and Vue 2 compatibility plugin...you can see from the doc that it is missing some features, probably because it was a "work in progress".
Another (sad/bad) detail you can see is that it was abandoned, so I [forked](https://github.com/Micene09/vc2c) it and added most of the the missing features (or at least the ones that I needed), that you can notice from the commits.
Also, I've added a new method to convert every file matched by an array of glob selectors.

I know guys, I know...this cannot be the real solution, neither for me and my company's projects, but I have tried it and it can be a good place to start the conversion of big projects, instead of convert all from scratch.
Just to be clear, here is the way I'm going to face it (for every project):
**Phase 1**: auto-convert (using vc2c)
**Phase 2**: manual refactor/fix errors to make it work

I hope I have helped someone with this comment.

**Comment by serebrov** [comment]
Actually both vue-class-component and vue-property-decorator already have release candidates with Vue 3 support:

* vue-class-component: 8.0.0-rc.1
* vue-property-decorator: 10.0.0-rc.3

It works well for me and I was able to switch Vue 2 app to Vue 3 (even without the migration build,`@vue/compat`, vue-class-component serves as a bridge between vue 2 and vue 3 here).
 
One catch was that the base `Vue` class for components needs to be imported from `vue-class-component`, not from the `vue-property-decorator` (as we had it before). It works in the actual app, but fails in tests (vue-test-utils), see #524.

It was still a lot of work, since I've also upgraded other related libs - `vue-router`, `vue-test-utils`, etc, but manageable and much faster than rewriting all components.

**Comment by JohnBernardsson** [comment]
> > We are already migrating every class-component to usual Object options API. Hard thing to do, but we needed to take some decision in order to update to Vue 3 ASAP...
> > I am a bit dissapointed though. Although I appreciate A LOT the work that has been done on this fantastic lib, I would have preferred that the communication about the future of it had been better. I waited months to take the decision pointed above, just because of this lack of clarity.
> 
> Why not migrating to the composition api? What was the reason for migrating from class components to the options API?

1. Well, Because we cannot simply stop the development while we migrate everything. We have 4 different projects of different sizes built with Vue 2. Our company (and our clients) demands functionality (which is understandable), and simply stopping everything for X number of weeks (months?) while devs perform the migration, overcomes whatever problems they found down the road, etc, is simply not an option. Migrating to the options API allows us to continue development new stuff in Vue 2 while at the same time we are preparing for a lot easier Vue 3 transition.

2. Also because the options API is a completely direct, no-brainer migration. The team still doesn't have real/professional experience with the composition API (neither do I), so doing the mentioned above, and at the same time trying to figure our what is the best way to transform our code to correctly use / better leverage the composition API... again is not an option for us. We will migrate to the options API and then either give a chance to the composition API in specific new functionalities, or try to do some refactor on specific pieces, and see how that goes, and what we learn in the process.

At least... that is the plan! :)

**Comment by charles-allen** [comment]
@JohnBernardsson - I'm not sure I understand your strategy. If you just want to get to Vue 3 (& not immediately to Composition), why don't you just leave your code as vue-class-component syntax?

1. Migrate to Vue 3 by adopting the latest VCC (minor syntax tweaks; much less effort than switching to OptionsAPI)
2. Migrate to CompositionAPI whenever you extend/refactor a component

**Comment by JohnBernardsson** [comment]
> @JohnBernardsson - I'm not sure I understand your strategy. If you just want to get to Vue 3 (& not immediately to Composition), why don't you leave your code as vue-class-component syntax?
> 
> 1. Migrate to Vue 3 by adopting the latest VCC (minor syntax tweaks; much less effort than switching to OptionsAPI)
> 2. Migrate to CompositionAPI whenever you extend/refactor a component
> 
> Is there some issue with Vue3/VCC that I missing? Stability?

Hey @charles-allen, some time ago I tried the latest, kind-of Vue-3-ready VCC, I don't remember right now what went wrong, but I had some issues with it. Maybe it was this, or it was when using it with vue-property-decorator (we are using them both) altogether, I don't remember clearly. There are some comments around about "gotchas" of people using those.

Also, those "latest" versions are "release candidates" at best. I cannot do a release of a webapp used by dozens of clients (a bunch of them really BIG companies) and thousands of users, with a "release candidate". I really need a production-ready state before going ahead with a migration.

**Comment by SHpiller887** [comment]
https://github.com/vuejs/vue-class-component/issues/406#issuecomment-916749627
This bug is not fixed. Extends from a component causes an error. Someone solved it?

**Comment by rdhelms** [comment]
I think @ruojianll 's vue-facing-decorator has a ton of potential...of all the alternatives to the original vue-class-component and vue-property-decorator configs that I've seen so far, that seems to work the best with our existing code with minimal changes. Has anyone else had a chance to check that out? I would love if we could come up with an official vue3 successor to these packages, since the maintainers here don't seem to be responding anymore

https://facing-dev.github.io/vue-facing-decorator/
https://github.com/vuejs/vue-class-component/issues/569#issuecomment-1145882492

**Comment by armano2** [comment]
> I think @ruojianll 's vue-facing-decorator has a ton of potential...

there is few thinks missing but we def can make it work, i will most likely spend some time to help folks over there to fix bugs

**Comment by zhennann** [comment]
Hello everyone, I have created a vue3 framework with ioc container, named as `Cabloy-Front`.

You can find it here: [https://github.com/cabloy/cabloy-front](https://github.com/cabloy/cabloy-front)

Do you agree with such a concept: Class should not be used in the `view layer`, but in the `business layer`.

Therefore, it is completely different from `vue-facind-decorator` or `vue-class-component`  that the Vue3 setup syntax sugar is still used to define vue components (such as props and emits of components), and `ioc container` and `class` are introduced only at the business level. In this way, combine the advantages of the two (function/class) to make the code more concise and elegant.

Cabloy-Front has a very important feature: With the support of ioc container, defining reactive states no longer needs `ref/reactive`. Without `ref`, naturally there is no need to write a lot of `ref.value`.

## Demonstration: no `ref/reactive`, no `ref.value`

[![Edit zhennann/cabloy-front-demo-codesandbox/main](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/p/github/zhennann/cabloy-front-demo-codesandbox2/main?embed=1&file=%2Fsrc%2Fsuite%2Fa-demo%2Fmodules%2Fa-demo%2Fsrc%2Fpage%2Fcounter%2Fcontroller.ts)

### 1. Define reactive state

Define a reactive variable `count` in the component and add two methods to modify its value

```typescript
export class MotherPageCounter {
  count: number = 0;

  inrement() {
    this.count++;
  }

  decrement() {
    this.count--;
  }
}
```

### 2. Use reactive state

Use `count` in render class

```typescript
export class RenderPageCounter {
  render() {
    return (
      <div>
        <div>count(ref): {this.count}</div>
        <button onClick={() => this.inrement()}>Inrement</button>
        <button onClick={() => this.decrement()}>Decrement</button>
      </div>
    );
  }
}
```

## Demonstration: dependency injection

[![Edit zhennann/cabloy-front-demo-codesandbox/main](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/p/github/zhennann/cabloy-front-demo-codesandbox2/main?embed=1&file=%2Fsrc%2Fsuite%2Fa-demo%2Fmodules%2Fa-demo%2Fsrc%2Fpage%2Fcounter2%2Fcontroller.ts)

### 1. Logic Reuse

Create a `Counter` Bean to implement the logic of `count`

```typescript
@Local()
export class Counter {
  count: number = 0;

  inrement() {
    this.count++;
  }

  decrement() {
    this.count--;
  }
}
```

### 2. Inject and use in a component

```typescript
export class MotherPageCounter {
  @Use()
  $$counter: Counter;
}
```

```typescript
export class RenderPageCounter {
  render() {
    return (
      <div>
        <div>count(ref): {this.$$counter.count}</div>
        <button onClick={() => this.$$counter.inrement()}>Inrement</button>
        <button onClick={() => this.$$counter.decrement()}>Decrement</button>
      </div>
    );
  }
}
```
